website/blog/buck2.md

```diff
@@ -563,3 +566,116 @@ I hope the current excitement could grow a community and ecosystem around Buck2
 
 In the meantime, I have created a repository to collect [all Buck2-related resources here](https://github.com/sluongng/awesome-buck2).
 If you are interested in Buck2, give it a look!
+
+## Update (2023-06-15)
+
+Shortly after this blog was released, [Tobias Werth](https://github.com/meisterT) from Bazel team pointed out to me
+that the Bazel setup was not the most optimized.
+
+In particular, instead of using the `tools` attribute in Bazel's `genrule`, I should have used `srcs` attribute instead.
+
+```diff
+--- a/BUILD.bazel
++++ b/BUILD.bazel
+@@ -26,28 +26,28 @@ SLEEP_CMD = "sleep $$((RANDOM % 5));"
+ [
+     genrule(
+         name = "z_{}".format(n),
+-        outs = ["z_{}.txt".format(n)],
+-        cmd = SLEEP_CMD + "cat $(location :x_{n}) $(location :y_{n}) > $@".format(n = n),
+-        tools = [
++        srcs = [
+             ":x_{}".format(n),
+             ":y_{}".format(n),
+         ],
++        outs = ["z_{}.txt".format(n)],
++        cmd = SLEEP_CMD + "cat $(location :x_{n}) $(location :y_{n}) > $@".format(n = n),
+     )
+     for n in SIZE_RANGE
+ ]
+
+ genrule(
+     name = "xyz",
++    srcs = [
++        ":z_{}".format(n)
++        for n in SIZE_RANGE
++    ],
+     outs = ["xyz.txt"],
+     # cat $(location :z_0) $(location :z_1) ... $(location :z_999) > $@,
+     cmd = SLEEP_CMD + "cat " + " ".join([
+         "$(location :z_{})".format(n)
+         for n in SIZE_RANGE
+     ]) + " > $@",
+-    tools = [
+-        ":z_{}".format(n)
+-        for n in SIZE_RANGE
+-    ],
+ )
+```
+
+Using `tools` would trigger Bazel to configure twice the number of actions underneath due to Bazel's special treament for tool dependencies.
+
+And fewer actions means faster build.
+For example, here is the new benchmark when build and test everything with Bazel.
+
+```bash
+> hyperfine --prepare 'bazel clean' --warmup 1 'bazel test --config=local //...'
+Benchmark 1: bazel test --config=local //...
+  Time (mean ± σ):     42.580 s ±  1.881 s    [User: 0.039 s, System: 0.048 s]
+  Range (min … max):   38.127 s … 45.369 s    10 runs
+```
+
+So I reran the benchmark and here are the new timings.
+
+| Tools       | Local Build           | RBE (with Remote Cache)   | RBE (without Remote Cache)   |
+|:------------|----------------------:|--------------------------:|-----------------------------:|
+| buck2 (old) | ` 1.175 s ±  2.183 s` |     ` 7.937 s ±  0.487 s` |       `203.431 s ± 29.882 s` |
+| bazel (old) | `65.631 s ±  0.460 s` |     ` 9.798 s ±  0.515 s` |       `106.062 s ± ‎ 3.106 s` |
+| bazel (new) | `42.580 s ±  1.881 s` |     `10.047 s ±  0.517 s` |       ` 73.009 s ± ‎ 2.975 s` |
```

Comment:
Mildly interesting: there are 2 "empty" Unicode characters here, right in front of `3.106 s` and `2.975 s` to help with the indentation on the blog.

Github does not render them but they are there.